WordDistanceInSpark.scala

```diff
@@ -1 +1,241 @@
+    
+    // Our input data is like a pseudo-timeseries, with word indices as time axis, 
+    // and only two possible values: word1 or word2.
+    // At this point we can still have repetitions of the same word.
+    // But a smallest distance can only happen when the signal transitions between hits of word1 and word2 or other other.
+    // Our interface allows us to not care about the indices of the transition itself, 
+    // only about the index-distance  between the words forming the transition
+    // Since we don't care about the transitions, automatic/manual stream fusion should in theory 
+    // allow to get rid of the transition-finding as a separate step.
+    // But given the time limitation, and because this kind of tweaking might be 
+    // much less important than other changes, I'm writing it here as the conceptual sequence of 
+    // first finding transitions, and then finding the smallest one.
+
+    // I wish I knew spark well enough already to be able to split the RDD of Hits in partitions
+    // which do not break a transition across partitions, in order to avoid a synchronization step.
+    // I immediately see two approaches that would work: 
+    // 1) If I could include the hit which happens to be part of a signal transition in both partitions, I'd be sure
+    // that no transition is broken up across partitions, and then the partitions 
+    // would only need to communicate back their winning candidate. seems fail-safe
+    // 2) If I could align the partition boundaries to only fall in the middle of hits of the same word
+    // we also would not have a broken transition.
+    // But if we only do option 2, our number of partitions would depend strongly on the input text.
+    
+    // Without a solution for avoiding the synchronization, I don't yet have a sense for how much of an improvement 
+    // it gives to do this part of the computation on the cluster instead of in the driver.
+    // For a non-pathological case, the filtering all words to just our word1 and word2 should already be 
+    // a huge shrink in dataset size.
+    
+    // but since there was a big warning against the trivial "make it run on spark" 
+    // implementation of immediately collecting, I'll try to write it in a spark-compatible style.
+    // 
+    // Stylewise, I'm unhappy because it becomes a bit car/cdr soup though, 
+    // which could be solved by defining case classes.
+
+    // I can't find a spark foldLeft which lets me fold Hits into SameWordIntervals.
+    // A verbose workaround but pretty foolproof solution would be to create a trait as union type.
+    // In this case I took the "wrap everything as a list" approach. Pretty ugly.
+    // I'm not sure whether defining a custom accumulator would help for making it less ugly 
+    // and for making it more performant.
+     
+    val reverseOrderedSameWordIntervals:List[SameWordInterval] = convertHitsToSameWordIntervals(orderedHits)
```

Comment:
I agree with you that this is bad style. The fact that the sort order is reversed should be clear from the method name, because that is a pretty significant side effect, and crucial for the further functioning of the code.

Looking back, I did not plan for the results to come back in reverse order when typing out the method call here. When I later adapted val sameWordIntervals to val reverseOrderedSameIntervals, I should have adjusted the method name as well.